выполнены домашние задания 7-1, 7-2, 7-3

wrhmmstr opened the pull request:
Выполнены домашние задания:
Тренировка #07-1: Вывод отсутствующего сообщения (null)
Тренировка #07-2: Вывод последовательности строковых сообщений в режиме прямой и обратной сортировки (ASC/DESC)
Тренировка #07-3: Вывод строковых сообщений без дублирования (DISTINCT)

Commit:
feat: обработка null значений

Добавлена обработка null значений в рамках домашнего задания Тренировка #07-1: Вывод отсутствующего сообщения (null)

Commit:
fix: поправил вывод сообщений при level=null

Поправил вывод сообщений при level=null, чтобы попасть в условие проверки "null-значения игнорируются, все остальные значения выводятся корректно"

Commit:
refactor: оптимизировал обработку null значений

Оптимизировал обработку null значений в рамках домашнего задания Тренировка #07-1: Вывод отсутствующего сообщения (null)

Commit:
feat: добавил перечислимый тип для сортировки

Добавил перечислимый тип для сортировки в рамках домашнего задания Тренировка #07-2: Вывод последовательности строковых сообщений в режиме прямой и обратной сортировки (ASC/DESC)

Commit:
feat: добавил возможность сортировки

Добавил возможность сортировки через перегруженный метод в рамках домашнего задания Тренировка #07-2: Вывод последовательности строковых сообщений в режиме прямой и обратной сортировки (ASC/DESC)

Commit:
fix: исправлена ASC сортировка

Исправлена ASC сортировка, добавил break в конструкцию switch

Commit:
feat: добавил возможность вывода без дублей

Добавил возможность вывода без дублей в новом перегруженном методе, но не реализовал поддержку сортировки из предыдущего перегруженного метода, в рамках домашнего задания Тренировка #07-3: Вывод строковых сообщений без дублирования (DISTINCT)

Commit:
fix: исправил вывод без дублей с обратной сортировкой

Исправил вывод без дублей с обратной сортировкой, в рамках домашнего задания Тренировка #07-3: Вывод строковых сообщений без дублирования (DISTINCT). Выглядит не очень красиво, буду думать как оптимизировать код.

Commit:
test: изменил тестовый пример

Изменил тестовый пример для проверки домашнего задания

Commit:
refactor: убрал лишние проверки на null

arsen-sepikhanov commented on src/main/java/com/tcs/edu/decorator/MessageService.java:
```diff
@@ -8,22 +10,155 @@
+            } else {
+                print(String.format("%s %s", decorate(message), messageToPage(messageCount)));
+            }
+        } else ;
```
Пустое else принято не писать вообще

wrhmmstr replied:
Исправлено

arsen-sepikhanov commented on src/main/java/com/tcs/edu/decorator/MessageService.java:
```diff
@@ -8,22 +10,155 @@
+     * @see MessageService Родительский класс
+     */
+    public static void processMessages(Severity level, MessageOrder order, Doubling doubling, String message, String... messages) {
+        switch (doubling) {
```
Что если doubling или order - null? Необходимо  как либо учесть эту ситуацию

wrhmmstr replied:
Исправлено

arsen-sepikhanov commented on src/main/java/com/tcs/edu/decorator/MessageService.java:
```diff
@@ -8,22 +10,155 @@
+     * @param messages Массив строк (String varargs) с сообщениями для декорирования
+     * @see MessageService Родительский класс
+     */
+    public static void processMessages(Severity level, MessageOrder order, String message, String... messages) {
```
Если воспользоваться перегрузкой аргументов, можно будет удалить копипаст, вызвав из этого метода соседний

Аналогично можно перегрузить все методы, кроме собственно самого большого, в котором реализована логика

wrhmmstr replied:
Исправлено

arsen-sepikhanov commented on src/main/java/com/tcs/edu/decorator/MessageService.java:
```diff
@@ -8,22 +10,155 @@
+                        printedMessages[printedWritedMessage] = message;
+                        for (int currentMessage = 0; currentMessage < messages.length; currentMessage++) {
+                            boolean isPrinted = false;
+                            for (int printedMessage = 0; printedMessage < printedMessages.length; printedMessage++) {
```
Здесь необязательно по всему массиву проходиться, ты ведь заранее знаешь, дальше какой позиции в этом массиве точно не будет элементов. Справедливо и для DESC ветки

wrhmmstr replied:
Исправлено

arsen-sepikhanov commented on src/main/java/com/tcs/edu/decorator/MessageService.java:
```diff
@@ -8,22 +10,155 @@
+                processMessages(level, order, message, messages);
+                break;
+            }
+            case DISTINCT: {
```
Ну и в целом стоит как то зарефакторить эту функцию, возможно разбить на несколько функций поменьше, устранить дублирование где можно

wrhmmstr replied:
Исправлено

Commit:
test: изменил тестовый пример

Изменил тестовый пример для проверки домашнего задания

Commit:
fix: исправил замечания от ревьюера

Исправил обработку doubling или order - null, избавился от дублирования где смог, порефакторил большую функцию вывода сообщений с doubling
